File: AEDS/AEDS2/src/exercicios/estruturas/lista/Lista.java
package exercicios.estruturas.lista;

import java.io.EOFException;

import exercicios.estruturas.Celula;

public class Lista {
  Celula inicio, fim;

  public Lista() {
    inicio = new Celula();
    fim = inicio;
  }

  public void mostrar() {
    Celula i = inicio.prox;
    while (i != null) {
      System.out.println(i.elemento);
      i = i.prox;
    }
  }

  public int tam() {
    int tam = 0;
    for (Celula i = inicio; i.prox != null; i = i.prox, tam++)
      ;
    return tam;
  }

  public void inserirNoInicio(int x) {
    Celula tmp = new Celula(x);
    tmp.prox = inicio.prox;
    inicio.prox = tmp;
    tmp = null;
  }

  public void inverter() {
    Celula i = inicio;
    Celula j = fim;
    System.out.println(j.elemento);
    Celula k = inicio.prox;
    for (; k.prox != j; k = k.prox) {
      System.out.println(k.elemento);
    }
    ;
    int e = i.elemento;
    i.elemento = j.elemento;
    j.elemento = e;
    i = i.prox;
    j = k;
  }

  public void inserir(int x, int pos) throws Exception {
    if (pos > this.tam() || pos < 0) {
      throw new Exception();
    }
    Celula tmp = inicio;
    for (int i = 0; i < pos; i++, tmp = tmp.prox) {

    }
    ;
    Celula nova = new Celula(x);
    nova.prox = tmp.prox;
    tmp.prox = nova;
    tmp = null;
  }

  public void inserirNoFim(int x) {
    Celula nova = new Celula(x);
    fim.prox = nova;
    fim = nova;
  }

  public int retirarDoInicio() {
    int e = inicio.prox.elemento;
    Celula tmp = inicio;
    inicio = inicio.prox;
    tmp.prox = null;
    tmp = null;
    return e;
  }

  public int remover(int pos) throws Exception {
    if (pos > this.tam()) {
      throw new Exception();
    }
    Celula tmp = inicio;
    for (int i = 0; i < pos; i++, tmp = tmp.prox)
      ;
    int e = tmp.prox.elemento;
    tmp.prox = tmp.prox.prox;
    tmp.prox = null;
    tmp = null;
    return e;
  }

  public int removerNoFim() {
    Celula i = inicio.prox;
    for (; i.prox != null; i = i.prox)
      ;
    int e = fim.elemento;
    fim = i;
    fim.prox = i = null;
    return e;
  }

  // Parte dos exercicios

  public int removeSegunda() throws Exception {
    if (inicio.prox != null || inicio.prox.prox != null) {
      throw new Exception();
    }
    Celula p = inicio.prox;
    Celula q = p.prox;
    int e = q.elemento;
    p.prox = q.prox;
    q.prox = null;
    q = null;
    return e;
  }

}

File: AEDS/AEDS2/src/exercicios/estruturas/lista/c/lista.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula
{
	int e;
	struct Celula *prox;
} Celula;

Celula *novaCelula(int x)
{
	Celula *nova = malloc(sizeof(Celula));
	nova->prox = NULL;
	nova->e = x;
	return nova;
};

int tam(Celula *cabeca)
{
	int tam = 1;
	Celula *atual = cabeca->prox;
	for (; atual != NULL; atual = atual->prox, tam++)
		;
	return tam;
}

void inserirNoInicio(Celula *cabeca, int x)
{
	Celula *nova = novaCelula(x);
	nova->prox = cabeca->prox;
	cabeca->prox = nova;
}

void inserir(Celula *cabeca, int x, int pos)
{
	if (pos > tam(cabeca))
	{
		printf("err");
		return;
	}

	Celula *atual = cabeca;
	Celula *nova = novaCelula(x);
	for (int i = 0; i < pos; atual = atual->prox, i++)
		;
	nova->prox = atual->prox;
	atual->prox = nova;
}

void inserirNoFim(Celula *cabeca, int x)
{
	Celula *nova = novaCelula(x);

	Celula *atual = cabeca;
	while (atual->prox != NULL)
	{
		atual = atual->prox;
	}
	atual->prox = nova;
};

int removerNoFim(Celula *cabeca)
{
	Celula *atual = cabeca;
	while (atual->prox->prox != NULL)
	{
		atual = atual->prox;
	}

	int e = atual->prox->e;
	Celula *ultimo = atual->prox;
	atual->prox = NULL;
	free(ultimo);
	return e;
};

int removerNoInicio(Celula *cabeca)
{
	Celula *primeiro = cabeca->prox;
	cabeca->prox = primeiro->prox;
	primeiro->prox = NULL;
	int e = primeiro->e;
	free(primeiro);
	return e;
}

int remover(Celula *cabeca, int pos)
{
	if (pos > tam(cabeca))
	{
		printf("err");
		return -1;
	}

	Celula *atual = cabeca;
	for (int i = 0; i < pos; atual = atual->prox, i++)
		;
	Celula *tmp = atual->prox;
	atual->prox = atual->prox->prox;
	tmp->prox = NULL;
	int e = tmp->e;
	free(tmp);
	return e;
}

void imprime(Celula *cabeca)
{
	Celula *atual = cabeca->prox;
	printf("[");
	for (; atual != NULL; atual = atual->prox)
	{
		printf("%i ", atual->e);
	}
	printf("]\n");
};

void limpar(Celula *cabeca)
{
	Celula *atual = cabeca->prox;
	Celula *libera;
	while (atual != NULL)
	{
		libera = atual;
		atual = atual->prox;
		free(libera);
	}
};

void inverteRec(Celula *i, Celula *j)
{
	if (i == j)
	{
		return;
	}
	int e = i->e;
	i->e = j->e;
	j->e = e;

	Celula *tmp = i;
	for (; tmp->prox != j; tmp = tmp->prox)
		;
	inverteRec(i->prox, tmp);
}

void inverte(Celula *cabeca)
{
	Celula *i = cabeca->prox;
	Celula *j = cabeca;
	for (; j->prox != NULL; j = j->prox)
		;
	inverteRec(i, j);
}

int main()
{
	Celula *cabeca = novaCelula(0);

	inserirNoFim(cabeca, 1);
	inserirNoFim(cabeca, 2);
	inserir(cabeca, 3, 2);

	imprime(cabeca);

	inverte(cabeca);
	imprime(cabeca);

	limpar(cabeca);
	return 0;
}
